Remotion/ObjectBinding/Web.ClientScript/Scripts/BocAutoCompleteReferenceValue.ts

```diff
@@ -0,0 +1,352 @@
+        },
+        formatItem: function (item: JQuery) //What we display on input box
+        {
+          var row = $('<li/>');
+
+          if (item.IconUrl != '')
+          {
+            var img = $('<img/>');
+            img.attr({ src: item.IconUrl, alt: '', 'aria-hidden': 'true' });
+            row.append($('<div/>').append(img));
+          }
+
+          var displayName = $('<span/>');
+          displayName.text(item.DisplayName);
+          row.append($('<div/>').append(displayName));
+
+          return {
+            html: row.html(),
+            class: null,
+            isAnnotation: item.IsAnnotation
+          };
+        },
+        formatMatch: function (item: BocAutoCompleteReferenceValue_Item) //The value used by the cache
+        {
+          return item.DisplayName;
+        },
+        handleRequestError: function (err: Sys.Net.WebServiceError)
+        {
+          SetError(resources.LoadDataFailedErrorMessage);
+        },
+        clearRequestError: function ()
+        {
+          ClearError();
+        }
+      }
+    ).invalidateResult(function (autoComplete: BocAutoCompleteReferenceValue) {
+      if (autoComplete._isInvalidated)
+        return;
+
+        autoComplete._isInvalidated = true;
+
+      hiddenField.val(nullValueString);
+      UpdateCommand(autoComplete, nullValueString);
+      //Do not fire change-event
+    }).updateResult(function (this: BocAutoCompleteReferenceValue, e: any, item: BocAutoCompleteReferenceValue_Item, out: any) //TODO what is this
+    {
+      try
+      {
+        var actualItem = item;
+
+        if (item.DisplayName.toLowerCase() == this._itemBackUp!.DisplayName.toLowerCase()
+          && (item.UniqueIdentifier == this._itemBackUp!.UniqueIdentifier || item.UniqueIdentifier == nullValueString))
+        {
+          if (item.UniqueIdentifier == nullValueString && this._itemBackUp!.UniqueIdentifier != nullValueString)
+            textbox.val(this._itemBackUp!.DisplayName); // fall back to the last confirmed user input to preserve correct casing
+          else
+            textbox.val(item.DisplayName); // keep the latest user input to preservce current casing
+          actualItem = this._itemBackUp!;
+        }
+
+        var hasChanged = this._itemBackUp != actualItem;
+
+        if (this._isInvalidated || hasChanged)
+        {
+          hiddenField.val(actualItem.UniqueIdentifier);
+          BackupItemData(this, actualItem.UniqueIdentifier, actualItem.DisplayName);
+
+          if (hasChanged)
+          {
+            UpdateCommand(this, actualItem.UniqueIdentifier);
+            this._commandBackUp = this._command;
+            hiddenField.trigger('change');
+          }
+          else
+          {
+            RestoreCommand(this);
+          }
+        }
+        out.Value = actualItem;
+      }
+      finally
+      {
+        this._isInvalidated = false;
+      }
+    });
+
+    function RestoreCommand(autocomplete: BocAutoCompleteReferenceValue)
```

Comment:
We should not modify the existing functions but instead use the `this` parameter. Since those functions also use the parameters of the function it gets a bit tricky since we cannot convert them to methods.

Instead we can convert the methods into variables and assign lambdas to them. e.g. `const RestoreCommand = () => { ... }`. This allows them to use `this` instead of introducing a new parameter. Since those methods are used above we need to moved them above their usage.